### ISSUE TITLE
Linux CLI Chat render broken? (Unable to load shared library 'libHarfBuzzSharp')

### ISSUE THREAD

**Issue by suvis1** [lay295/TwitchDownloader#218]
Hi!

I recently built TwitchDownloader again on my ARM64 linux box, and chatrender fails with :

`Unhandled exception. System.AggregateException: One or more errors occurred. (Unable to load shared library 'libHarfBuzzSharp' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: liblibHarfBuzzSharp: cannot open shared object file: No such file or directory)`

**Comment by suvis1** [comment]
I managed to fix it by including `HarfBuzzSharp.NativeAssets.Linux` into TwitchDownloaderCore, but now it fails with:

```
Unhandled exception. System.AggregateException: One or more errors occurred. (Value cannot be null. (Parameter 'typeface'))
 ---> System.ArgumentNullException: Value cannot be null. (Parameter 'typeface')
```

I am not a coder so im kinda stumped now :S maybe this has something to do with it? :) 

https://github.com/mono/SkiaSharp/issues/1058

**Comment by lay295** [comment]
Whoops I'll be sure to include that nuget package in my next commit.

As for the typeface issue, in the CLI it defaults the font as "Arial" which might not be installed. You can install ttf-mscorefonts-installer and refresh the font cache and it should hopefully work, or specify a font that's installed on the system.

If neither of those work, I can try and look into it tomorrow.

**Comment by suvis1** [comment]
I tried installing `ttf-mscorefonts-installer` and refreshing font cache with `sudo fc-cache -f -v` . I also tried fonts with name (like DejaVu Sans, sans, etc) and even with full paths to .ttf files. Nothing worked :(  Heres a full error log for you:

```
Unhandled exception. System.AggregateException: One or more errors occurred. (Value cannot be null. (Parameter 'typeface'))
 ---> System.ArgumentNullException: Value cannot be null. (Parameter 'typeface')
   at SkiaSharp.HarfBuzz.SKShaper..ctor(SKTypeface typeface)
   at SkiaSharp.HarfBuzz.CanvasExtensions.DrawShapedText(SKCanvas canvas, String text, Single x, Single y, SKPaint paint)
   at TwitchDownloaderCore.ChatRenderer.DrawText(SKBitmap sectionImage, String message, SKPaint messageFont, List`1 imageList, ChatRenderOptions renderOptions, List`1 currentGifEmotes, Size canvasSize, Point& drawPos, Boolean padding, Int32 default_x) in /home/odroid/TwitchDownloader/TwitchDownloaderCore/ChatRenderer.cs:line 911
   at TwitchDownloaderCore.ChatRenderer.DrawMessage(SKBitmap sectionImage, List`1 imageList, ChatRenderOptions renderOptions, List`1 currentGifEmotes, SKPaint messageFont, Dictionary`2 emojiCache, List`1 chatEmotes, List`1 thirdPartyEmotes, List`1 cheerEmotes, Comment comment, Size canvasSize, Point& drawPos, Int32& default_x, List`1 emoteList, List`1 emotePositionList) in /home/odroid/TwitchDownloader/TwitchDownloaderCore/ChatRenderer.cs:line 783
   at TwitchDownloaderCore.ChatRenderer.<>c__DisplayClass6_1.<RenderVideoAsync>b__0() in /home/odroid/TwitchDownloader/TwitchDownloaderCore/ChatRenderer.cs:line 119
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location where exception was thrown ---
   at TwitchDownloaderCore.ChatRenderer.RenderVideoAsync(IProgress`1 progress, CancellationToken cancellationToken) in /home/odroid/TwitchDownloader/TwitchDownloaderCore/ChatRenderer.cs:line 70
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at TwitchDownloaderCLI.Program.RenderChat(Options inputOptions) in /home/odroid/TwitchDownloader/TwitchDownloaderCLI/Program.cs:line 250
   at TwitchDownloaderCLI.Program.Main(String[] args) in /home/odroid/TwitchDownloader/TwitchDownloaderCLI/Program.cs:line 66
Aborted (core dumped)
```

**Comment by lay295** [comment]
Could you send me the chat json you're trying to render?

**Comment by suvis1** [comment]
[tempchat.zip](https://github.com/lay295/TwitchDownloader/files/7151003/tempchat.zip)

heres the args im using:

```
-m ChatRender \
-i /home/odroid/vod_test/tempchat.json \
-h 260 -w 270 \
--framerate 10 \
--update-rate 0 \
--font-size 12 \
--background-color "#00000000" \
--generate-mask \
--outline \
-o /home/odroid/vod_test/temprender.mp4
```

**Comment by lay295** [comment]
OK should be fixed now in the newest commit. I've been doing a few commits but haven't tested too much and was going to before I put out a new release.

When I changed from [DrawText](https://github.com/lay295/TwitchDownloader/commit/48b0a162e8b97ea6d5bba7f950a99d6068b4b766#diff-ca35d49087eef863cf91c18f7d2fe213631ffee6a5994e72d1a703348b37d734L802) to DrawShapedText to support RTL languages such as arabic, I guess DrawShapedText doesn't like null typefaces. It got a null typeface because my font fallback failed, and with DrawText it used to just draw a little box showing that it couldn't find a font, but I guess DrawShapedText throws an error instead. I've just made it use DrawText in DrawShapedText fails.

![image](https://user-images.githubusercontent.com/1060681/133023971-32b475b4-3414-4b12-aded-8ee8ffc5eb5c.png)

It was caused by someone using an [invisible character](https://www.compart.com/en/unicode/U+E0000), I'm guessing to send the same message within 30 seconds the user just appended an invisible character. Anyways give the new commit a build and lemme know if it's still broken :)
